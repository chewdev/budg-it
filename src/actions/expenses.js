import {
  database,
  ref,
  get,
  child,
  push,
  set,
  remove,
  update,
} from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const userExpensesRef = ref(database, `users/${uid}/expenses`);
    const newExpenseRef = push(userExpensesRef);

    return set(newExpenseRef, expense).then((ref) => {
      dispatch(
        addExpense({
          id: newExpenseRef.key,
          ...expense,
        })
      );
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenseToRemoveRef = ref(database, `users/${uid}/expenses/${id}`);
    return remove(expenseToRemoveRef).then(() => {
      dispatch(
        removeExpense({
          id,
        })
      );
    });
  };
};

export const removeExpensesChosen = (ids = []) => ({
  type: "REMOVE_EXPENSES_CHOSEN",
  ids,
});

export const startRemoveExpensesChosen = (ids = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    ids.forEach((id) => {
      const expenseToRemoveRef = ref(database, `users/${uid}/expenses/${id}`);
      return remove(expenseToRemoveRef).then(() => {
        dispatch(removeExpense({ id }));
      });
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenseToUpdateRef = ref(database, `users/${uid}/expenses/${id}`);
    return update(expenseToUpdateRef, updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return get(child(ref(database), `users/${uid}/expenses`)).then(
      (snapshot) => {
        const expenses = [];

        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      }
    );
  };
};
