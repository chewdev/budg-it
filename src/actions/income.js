import {
  database,
  ref,
  get,
  child,
  push,
  set,
  remove,
  update,
} from "../firebase/firebase";

// ADD_INCOME
export const addIncome = (income) => ({
  type: "ADD_INCOME",
  income,
});

export const startAddIncome = (incomeData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = incomeData;
    const income = { description, note, amount, createdAt };
    const userIncomesRef = ref(database, `users/${uid}/incomes`);
    const newIncomeRef = push(userIncomesRef);

    return set(newIncomeRef, income).then((ref) => {
      dispatch(
        addIncome({
          id: newIncomeRef.key,
          ...income,
        })
      );
    });
  };
};

// REMOVE_INCOME
export const removeIncome = ({ id } = {}) => ({
  type: "REMOVE_INCOME",
  id,
});

export const startRemoveIncome = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const incomeToRemoveRef = ref(database, `users/${uid}/incomes/${id}`);
    return remove(incomeToRemoveRef).then(() => {
      dispatch(
        removeIncome({
          id,
        })
      );
    });
  };
};

export const removeIncomesChosen = (ids = []) => ({
  type: "REMOVE_INCOMES_CHOSEN",
  ids,
});

export const startRemoveIncomesChosen = (ids = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    ids.forEach((id) => {
      const incomeToRemoveRef = ref(database, `users/${uid}/incomes/${id}`);
      return remove(incomeToRemoveRef).then(() => {
        dispatch(removeIncome({ id }));
      });
    });
  };
};

// EDIT_INCOME
export const editIncome = (id, updates) => ({
  type: "EDIT_INCOME",
  id,
  updates,
});

export const startEditIncome = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const incomeToUpdateRef = ref(database, `users/${uid}/incomes/${id}`);
    return update(incomeToUpdateRef, updates).then(() => {
      dispatch(editIncome(id, updates));
    });
  };
};

// SET_INCOMES
export const setIncomes = (incomes) => ({
  type: "SET_INCOMES",
  incomes,
});

export const startSetIncomes = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return get(child(ref(database), `users/${uid}/incomes`)).then(
      (snapshot) => {
        const incomes = [];
        snapshot.forEach((childSnapshot) => {
          incomes.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setIncomes(incomes));
      }
    );
  };
};
