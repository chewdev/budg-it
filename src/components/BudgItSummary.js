import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectIncome from '../selectors/income';
import selectTotal from '../selectors/total';

export const BudgItSummary = (props) => {
    const expenseWord  = props.expenses.length === 1 ? 'expense ' : 'expenses ';
    const incomeWord = props.income.length === 1 ? 'income ' : 'incomes ';
    const budgitTotal = selectTotal(props.expenses, props.income);
    return (
    <div className={budgitTotal >= 0 ? "page-header" : "page-header page-header--bg-red"}>
        <div className="content-container">
            <h2 className="page-header__budgit">
                Budget: <span className={budgitTotal === 0 ? "page-header__brown" : budgitTotal > 0 ? "page-header__main-green" : "page-header__red"}>{numeral(budgitTotal).format('$0,0.00')}</span>
            </h2>
            <div className="page-header__sub-summary">
                <h3 className="page-header__expenses">
                    Viewing <span className="page-header__expense-number">{props.expenses.length}</span> {expenseWord}
                    totaling: <span className="page-header__expense-total">{numeral(selectTotal(props.expenses)).format('$0,0.00')}</span>
                </h3>
                <h3 className="page-header__incomes">
                    Viewing <span className="page-header__income-number">{props.income.length}</span> {incomeWord}
                    totaling: <span className="page-header__income-total">{numeral(selectTotal(props.income)).format('$0,0.00')}</span>
                </h3>
            </div>
            <div className="page-header__sub-summary">
                <Link to='/create/expense' className="btn btn--expense">Add Expense</Link>
                <Link to='/create/income' className="btn btn--income">Add Income</Link>
            </div>
        </div>
    </div>
)};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters),
        income: selectIncome(state.income, state.filters)
    };
};

export default connect(mapStateToProps)(BudgItSummary);