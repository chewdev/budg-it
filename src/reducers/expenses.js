// Expenses Reducer

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'REMOVE_EXPENSES_CHOSEN':
            action.ids.forEach((id) => {
            state = state.filter((expense) => expense.id !== id);
            });
            return state;
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        case 'SET_EXPENSES':
            return action.expenses;
        case 'REMOVE_ALL':
            return [];
        default:
            return state;
    }
};