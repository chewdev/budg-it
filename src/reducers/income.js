// Income Reducer

const incomeReducerDefaultState = [];

export default (state = incomeReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_INCOME":
      return [...state, action.income];
    case "REMOVE_INCOME":
      return state.filter((income) => income.id !== action.id);
    case "REMOVE_INCOMES_CHOSEN":
      action.ids.forEach((id) => {
        state = state.filter((income) => income.id !== id);
      });
      return state;
    case "EDIT_INCOME":
      return state.map((income) => {
        if (income.id === action.id) {
          return {
            ...income,
            ...action.updates,
          };
        } else {
          return income;
        }
      });
    case "SET_INCOMES":
      return action.incomes;
    default:
      return state;
  }
};
