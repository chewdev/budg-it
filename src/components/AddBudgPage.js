import React from 'react';
import BudgForm from './BudgForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddBudgPage extends React.Component {
	onSubmit= (expense) => {
		// props.dispatch(addExpense(expense));
		this.props.startAddExpense(expense);
		this.props.history.push('/');
	};
	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<BudgForm 
					buttonText="Add Expense"
					onSubmit={this.onSubmit}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({startAddExpense: (expense) => dispatch(startAddExpense(expense))});
	

export default connect(undefined, mapDispatchToProps)(AddBudgPage);