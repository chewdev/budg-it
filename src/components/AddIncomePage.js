import React from 'react';
import BudgForm from './BudgForm';
import { connect } from 'react-redux';
import { startAddIncome } from '../actions/income';

export class AddIncomePage extends React.Component {
	onSubmit = (income) => {
		this.props.startAddIncome(income);
		this.props.history.push('/');
	};
	 render() { 
		return (
			<div>
				<h1>Add Income</h1>
				<BudgForm 
					buttonText="Add Income"
					onSubmit={this.onSubmit}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({startAddIncome: (income) => dispatch(startAddIncome(income))});

export default connect(undefined, mapDispatchToProps)(AddIncomePage);