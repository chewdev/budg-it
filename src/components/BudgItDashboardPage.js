import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import BudgItSummary from './BudgItSummary';
import IncomeList from './IncomeList';

const BudgItDashboardPage = () => (
	<div>
		<BudgItSummary />
		<ExpenseListFilters />
		<ExpenseList />
		<IncomeList />

	</div>
);

export default BudgItDashboardPage;