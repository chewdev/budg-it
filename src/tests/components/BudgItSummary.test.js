import React from 'react';
import { shallow } from 'enzyme';
import { BudgItSummary } from '../../components/BudgItSummary';
import expenses  from '../fixtures/expenses';
import income from '../fixtures/income';

test('should correctly render BudgItSummary with expenses and income', () => {
    const wrapper = shallow(<BudgItSummary expenses={expenses} income={income}  />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render BudgItSummary with expenses and no income', () => {
    const wrapper = shallow(<BudgItSummary expenses={expenses} income={[]}  />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render BudgItSummary with income and no expenses', () => {
    const wrapper = shallow(<BudgItSummary expenses={[]} income={income}  />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render BudgItSummary with no expenses and no income', () => {
    const wrapper = shallow(<BudgItSummary expenses={[]} income={[]}  />);
    expect(wrapper).toMatchSnapshot();
});