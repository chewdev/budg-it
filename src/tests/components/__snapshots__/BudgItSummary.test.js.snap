// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render BudgItSummary with expenses and income 1`] = `
<div>
  <h2>
    Budg-It: 
    $1,019.05
  </h2>
  <h3>
    Viewing 
    3
     
    expenses 
    totaling: 
    $124.85
  </h3>
  <h3>
    Viewing 
    3
     
    incomes 
    totaling: 
    $1,143.90
  </h3>
</div>
`;

exports[`should correctly render BudgItSummary with expenses and no income 1`] = `
<div>
  <h2>
    Budg-It: 
    -$124.85
  </h2>
  <h3>
    Viewing 
    3
     
    expenses 
    totaling: 
    $124.85
  </h3>
  <h3>
    Viewing 
    0
     
    incomes 
    totaling: 
    $0.00
  </h3>
</div>
`;

exports[`should correctly render BudgItSummary with income and no expenses 1`] = `
<div>
  <h2>
    Budg-It: 
    $1,143.90
  </h2>
  <h3>
    Viewing 
    0
     
    expenses 
    totaling: 
    $0.00
  </h3>
  <h3>
    Viewing 
    3
     
    incomes 
    totaling: 
    $1,143.90
  </h3>
</div>
`;

exports[`should correctly render BudgItSummary with no expenses and no income 1`] = `
<div>
  <h2>
    Budg-It: 
    $0.00
  </h2>
  <h3>
    Viewing 
    0
     
    expenses 
    totaling: 
    $0.00
  </h3>
  <h3>
    Viewing 
    0
     
    incomes 
    totaling: 
    $0.00
  </h3>
</div>
`;
